### PR TITLE
Check appservice user interest against the local users instead of all users (`get_users_in_room` mis-use)

### PR DESCRIPTION
**Pull request by MadLittleMods**
Check appservice user interest against the local users instead of all users (`get_users_in_room` mis-use).

`get_local_users_in_room` is way more performant since it looks at a single table (`local_current_membership`) and is looking through way less data since it only worries about the local users in the room instead of everyone in the room across the federation.

Fix https://github.com/matrix-org/synapse/issues/13942
(probably fixes, we don't know for sure but it's our hunch atm)

Related to https://github.com/matrix-org/matrix-spec/issues/1272

Related to the following PRs where we also cleaned up some `get_users_in_room` mis-uses:

 - https://github.com/matrix-org/synapse/pull/13605
 - https://github.com/matrix-org/synapse/pull/13608
 - https://github.com/matrix-org/synapse/pull/13606
 - https://github.com/matrix-org/synapse/pull/13960

See https://github.com/matrix-org/synapse/pull/13575#discussion_r953023755 for the original exploration around finding `get_users_in_room` mis-uses.

### Problem

https://github.com/matrix-org/synapse/pull/13575 made the `get_users_in_room` SQL query a little more complicated by [adding an extra join statement to the query](https://github.com/matrix-org/synapse/pull/13575#discussion_r952740418) which means it takes more time.

This extra database time can show up as a hot-spot for homeservers that make heavy use of application services and bridged users (like [Element One](https://element.io/element-one)), see https://github.com/matrix-org/synapse/issues/13942

### Pull Request Checklist

* [x] Pull request is based on the develop branch
* [x] Pull request includes a [changelog file](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#changelog). The entry should:
  - Be a short description of your change which makes sense to users. "Fixed a bug that prevented receiving messages from other servers." instead of "Moved X method from `EventStore` to `EventWorkerStore`.".
  - Use markdown where necessary, mostly for `code blocks`.
  - End with either a period (.) or an exclamation mark (!).
  - Start with a capital letter.
  - Feel free to credit yourself, by adding a sentence "Contributed by @github_username." or "Contributed by [Your Name]." to the end of the entry.
* [ ] ~~Pull request includes a [sign off](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#sign-off)~~
* [x] [Code style](https://matrix-org.github.io/synapse/latest/code_style.html) is correct
  (run the [linters](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#run-the-linters))

### CODE REVIEW

File: synapse/appservice/__init__.py [hunk 1]
```diff
@@ -172,15 +172,11 @@ async def _matches_user_in_member_list(
...
+        app_service_users_in_room = await store.get_app_service_users_in_room(
+            room_id, self, on_invalidate=cache_context.invalidate
         )
 
-        # check joined member events
-        for user_id in member_list:
-            if self.is_interested_in_user(user_id):
-                return True
-        return False
+        return len(app_service_users_in_room) > 0
```
Comment (MadLittleMods) [inline synapse/appservice/__init__.py:179]: As [@clokep mentioned](https://matrix.to/#/!SGNQGPGUwtcPBUotTL:matrix.org/$feRc-xR0MZ5e0sZjMrJIJgwFIJtIbhrK9zL7zHsnIKM?via=jki.re&via=matrix.org&via=element.io), `_matches_user_in_member_list` and `get_app_service_users_in_room` are suspiciously almost the same so we could re-use one in the other.

This technically takes more time the first time we hit this spot because before we would bail early as soon as we found one user that the appservice was interested in. But now we're at least re-using the `get_app_service_users_in_room` function which is also cached.

The only problem is that it looks like `get_app_service_users_in_room` is only used elsewhere in `_compute_msc3202_otk_counts_and_fallback_keys` so maybe it's not that common to use the cache. We're doing more work and barely using it later 🤔

File: synapse/handlers/room_member.py [hunk 2]
```diff
@@ -1150,7 +1150,7 @@ async def transfer_room_state_on_room_upgrade(
         logger.info("Transferring room state from %s to %s", old_room_id, room_id)
 
         # Find all local users that were in the old room and copy over each user's state
-        users = await self.store.get_users_in_room(old_room_id)
+        users = await self.store.get_local_users_in_room(old_room_id)
```
Comment (MadLittleMods) [inline synapse/handlers/room_member.py:1153]: TODO: Split this out to another PR

File: synapse/handlers/user_directory.py [hunk 3]
```diff
@@ -392,6 +392,7 @@ async def _track_user_joined_room(self, room_id: str, user_id: str) -> None:
         if is_public:
             await self.store.add_users_in_public_rooms(room_id, (user_id,))
         else:
+            # TODO: get_local_users_in_room here
```
Comment (MadLittleMods) [inline synapse/handlers/user_directory.py:395]: TODO: Split this out to another PR

File: synapse/handlers/room_member.py (hunk 2 above)
Comment (MadLittleMods) [reply synapse/handlers/room_member.py:1153]: Split out to https://github.com/matrix-org/synapse/pull/13960

File: synapse/handlers/user_directory.py (hunk 3 above)
Comment (MadLittleMods) [reply synapse/handlers/user_directory.py:395]: The logic is a bit confusing and subtle here but we do actually use remote users here,

https://github.com/matrix-org/synapse/blob/6f0c3e669da458e838e7b4b165a13e8a5312d6d0/synapse/handlers/user_directory.py#L407-L410

Refactored the code a bit here to make this fact more clear: https://github.com/matrix-org/synapse/pull/13966

File: synapse/appservice/__init__.py (hunk 1 above)
Comment (MadLittleMods) [reply synapse/appservice/__init__.py]: Reverted back to having them separate but added a comment about a future possibility of re-using when `get_app_service_users_in_room` is used more so the juice is worth the squeeze.

File: synapse/appservice/__init__.py [hunk 4]
```diff
@@ -172,7 +172,18 @@ async def _matches_user_in_member_list(
...
+        # service can only act on behalf of users of the server it's on.
+        #
+        # In the future, we can consider re-using
+        # `store.get_app_service_users_in_room` which is very similar to this
+        # function but has a slightly worse performance than this because we
+        # have an early escape-hatch if we find a single user that the
+        # appservice is interested in. The juice would be worth the squeeze if
+        # `store.get_app_service_users_in_room` was used in more places besides
+        # an experimental MSC. But for now we can avoid doing more work and
+        # barely using it later.
```
Comment (MadLittleMods) [inline synapse/appservice/__init__.py:185]: For context on this comment, see https://github.com/matrix-org/synapse/pull/13958#discussion_r984074247

File: synapse/storage/databases/main/roommember.py [hunk 5]
```diff
@@ -152,13 +152,18 @@ async def get_users_in_room(self, room_id: str) -> List[str]:
...
+
+        Will return inaccurate results for rooms with partial state, since the
+        state for the forward extremities of those rooms will exclude most
+        members. We may also calculate room state incorrectly for such rooms and
+        believe that a member is or is not in the room when the opposite is
+        true.
+
+        Note: If you only care about users in the room local to the homeserver,
+        use `get_local_users_in_room(...)` instead which will be more
+        performant.
```
Comment (MadLittleMods) [inline synapse/storage/databases/main/roommember.py:166]: Only change is this last paragraph to point prospective users of `get_users_in_room` to consider using `get_local_users_in_room` instead if it fits their use case.

The rest of the changes is auto-wrapping.

File: changelog.d/13958.bugfix [hunk 6]
```diff
@@ -0,0 +1 @@
+Fix performance bottleneck with heavy appservice and bridged users in Synapse by checking appservice user interest against the local users instead of all users.
```
Comment (MadLittleMods) [inline changelog.d/13958.bugfix:1]: For anyone curious, there might be a few more `get_users_in_room` mis-uses in the `sync` and `presence` code for someone who has more domain knowledge on what those things should exactly do.

File: synapse/storage/databases/main/roommember.py (hunk 5 above)
Comment (squahtx) [reply synapse/storage/databases/main/roommember.py:166]: I think we don't need to rewrap these lines? We've set up our linters to 88 columns, instead of 80.

File: synapse/appservice/__init__.py [hunk 7]
```diff
@@ -172,12 +172,23 @@ async def _matches_user_in_member_list(
...
+        local_user_ids = await store.get_local_users_in_room(
```
Comment (squahtx) [inline synapse/appservice/__init__.py:186]: This changes the behavior of `is_interested_in_room` slightly. Previously, we could match remote users when the appservice's user regex didn't include a hostname (all of our test cases and examples in the documentation don't). Was that just a bug?

File: synapse/appservice/__init__.py (hunk 1 above)
Comment (clokep) [reply synapse/appservice/__init__.py]: Oh the bailing early is a good call that this likely doesn't make sense. 👍

File: synapse/appservice/__init__.py (hunk 7 above)
Comment (clokep) [reply synapse/appservice/__init__.py:186]: > Was that just a bug?

I don't see how an appservice could control remote users, so I think so...but might be worth asking someone with more info?

Comment (anoadragon453) [reply synapse/appservice/__init__.py:186]: It's not really a question of controlling the users, but more of whether the application service wants to receive events and updates related to those users, which an application service may well want to do for events coming from non-local users.

So I think this is a change we wouldn't want to go ahead with.

File: synapse/storage/databases/main/roommember.py (hunk 5 above)
Comment (MadLittleMods) [reply synapse/storage/databases/main/roommember.py:166]: Ahh, thanks @squahtx, adjusted the wrapping in my text editor 🙂

File: synapse/appservice/__init__.py (hunk 7 above)
Comment (MadLittleMods) [reply synapse/appservice/__init__.py:186]: Interesting! I hadn't considered this use case.

For my own reference, the spec doesn't mention any restriction either: https://spec.matrix.org/v1.1/application-service-api/#registration

I've updated the comments to clarify local and remote so it's obvious if this is attempted again. We do have tests around this kinda but because they were just mocking `get_users_in_room`, it happily passed when I updated it to `get_local_users_in_room` as well. I've added some actual homeserver tests and verified that they fail if we use `get_local_users_in_room` here ✅

Comment (MadLittleMods) [reply synapse/appservice/__init__.py:186]: Created https://github.com/matrix-org/synapse/pull/14000 to introduce these clarification changes with so we can leave the diff in this PR for reference of what we didn't want to do.

**Comment by MadLittleMods** [comment]
Thanks for the review @squahtx, @clokep, and @anoadragon453 🐎

File: synapse/appservice/__init__.py (hunk 7 above)
Comment (MadLittleMods) [reply synapse/appservice/__init__.py:187]: The conclusion from https://github.com/matrix-org/synapse/pull/14000#issuecomment-1266552433 is that we don't actually want to support application services being interested in remote users. Thanks for bring it up @erikjohnston and @anoadragon453 for following up in #matrix-spec to clarify it with everyone,

> As @Half-Shot put it, defining remote users is merely a footgun for which an Application Service may assume that it'll receive all events sent by that user, even though it will only receive events in rooms that are shared with a local user.

The path forward is this PR again:

> So by that logic, I'm fine with us defining (and limiting in the code) user namespaces to only cover local users.

File: tests/handlers/test_appservice.py [hunk 8]
```diff
@@ -412,6 +413,100 @@ def prepare(self, reactor, clock, hs):
...
+            ("@remote_as_user:remote", False),
+        ]
+    )
+    def test_match_interesting_room_members(
+        self, interesting_user: str, should_notify: bool
+    ):
+        """
+        Test to make sure that a interesting user (local or remote) in the room is
+        notified as expected when someone else in the room sends a message.
+        """
```
Comment (MadLittleMods) [inline tests/handlers/test_appservice.py:446]: A note to previous reviewers, this parameterized test is new!

File: synapse/appservice/__init__.py (hunk 7 above)
Comment (turt2live) [reply synapse/appservice/__init__.py:187]: ftr, such a change would require an MSC I believe - the spec considers it a feature to be able to listen to register remote users in the namespace.

Comment (anoadragon453) [reply synapse/appservice/__init__.py:187]: That being said, the changes here are valid - simply only returning events targeted at local users covers the same use cases.

Comment (turt2live) [reply synapse/appservice/__init__.py:187]: I'm jumping into this about a week late, but if Synapse doesn't respect `@.*` out of the box then it'd be in violation of the spec - it sounds like this change *causes* such a case.

Comment (turt2live) [reply synapse/appservice/__init__.py:187]: I'm a bit concerned I might not fully understand the impact of the change here - from what I've been reading, it'd be a major spec violation for this to land, though I would be surprised if that were the actual case here. 

Is this literally just treating `@namespace.*` as `@namespace.*:localdomain`, or is there something else going on?

Comment (MadLittleMods) [reply synapse/appservice/__init__.py:187]: @turt2live This PR makes it so `@namespace.*` is only tried against the local users of the server.

Previously, we would try applying the regex to everyone in the room.

Comment (turt2live) [reply synapse/appservice/__init__.py:187]: That would be a major spec violation indeed, which makes this PR not safe to land (with my SCT hat on). We would need *at least* an MSC, which the Synapse team might impose additional restrictions upon (config requirements, etc). 

The full line of investigation on the spec side was to answer the question whether `@namespace.*` actually does match all users regardless of homeserver or if the spec was permissive to restricting to just the local homeserver. In searching the old matrix-doc repo (where the spec & proposals used to live as one, now just proposals), https://github.com/matrix-org/matrix-spec-proposals/issues/1307 came up as part of the r0 effort in 2018/2019, where we attempted to declare the spec out of "beta" and as production-quality. That issue clearly tries to communicate that it is in fact all users regardless of homeserver, and the spec PR reinforces it. What was clearly lacking (in hindsight) is that the spec phrasing itself was not any more helpful as it did not account for the `/user` endpoint or really anything else to do with namespaces. From a spec perspective, this is a regrettable situation but would mean that the spec *intended* to say the namespace can make an appservice interested in remote users, though there's obviously no ability for the server to call `/user` on remote users (it's not like the appservice can create them). Thus, it's specified as such and a clarification PR to solidify that state would be appreciated.

An MSC which changes that would also be good, as clearly it's a great way to cause problems in various ways.

All told, this is a good reminder that the appservice spec is still largely in its pre-2018 state (it didn't get a whole lot of editing for r0) and didn't get the attention it needed. Going back, I probably would have held r0 back in favour of figuring out issues like this and namespaces in general (there's plenty of cases for "wtf, how does this even work"). We're too far along now to go back like that though, making the only option persevering and fixing it :)

(I don't think it'd be a very complicated MSC, which also means it could go through the process relatively quickly - not sure how much this helps with this PR's timelines, though)

Comment (MadLittleMods) [reply synapse/appservice/__init__.py:187]: @turt2live I feel like the [spec issue](https://github.com/matrix-org/matrix-spec-proposals/issues/1307) and PR doesn't clarify as strongly as you're deriving from it.

> you don't just match on the event's sender for the "users" namespace, you also send all events from a room if a user that matches is in that room.

This just talks about the behavior of matching room members and not just event senders, nothing at all clarifying which room members (remote and/or local). I think it could be assumed either way and makes more sense to assume only local from the whole discussion we're having now and our desire to clarify that it should be local only.

Comment (MadLittleMods) [reply synapse/appservice/__init__.py:187]: > though the historical context (which may be lost to time) is still very prevelant here.

Is there something specific to point to here? Or just from your memory?

File: synapse/appservice/__init__.py [hunk 9]
```diff
@@ -172,12 +172,24 @@ async def _matches_user_in_member_list(
         Returns:
             True if this service would like to know about this room.
         """
-        member_list = await store.get_users_in_room(
+        # We can use `get_local_users_in_room(...)` here because an application service
+        # can only be interested in local users of the server it's on (ignore any remote
+        # users that might match the user namespace regex).
```
Comment (MadLittleMods) [inline synapse/appservice/__init__.py:177]: We have to rip the band-aid off at some point whether it before or after [MSC3905](https://github.com/matrix-org/matrix-spec-proposals/pull/3905) merges. People relying on the old behavior will probably have to find out the hard way that it doesn't work anymore. I can't think of a way to add a warning for this scenario that wouldn't generate a ton of noise (false-positives) 🤔

We could detect a very specific case where someone explicitly defines the `@.*:hostname` part of the regex that doesn't match the server hostname.

File: synapse/appservice/__init__.py (hunk 7 above)
Comment (MadLittleMods) [reply synapse/appservice/__init__.py:187]: Created a MSC for this: [MSC3905](https://github.com/matrix-org/matrix-spec-proposals/pull/3905)


---

And prodding this point in case we can merge earlier,

> That being said, the changes here are valid - simply only returning events targeted at local users covers the same use cases.

I tend to agree with @anoadragon453 here (I think we agree 😬). The use cases of seeing events from remote users (like moderation) is already supported by matching the entire room with the `rooms` namespace regex so this PR should be ok to merge.

We have to rip the band-aid off at some point whether it before or after https://github.com/matrix-org/matrix-spec-proposals/pull/3905 merges. Maybe there is a way to detect people relying on the old behavior? -> https://github.com/matrix-org/synapse/pull/13958#discussion_r992977143

Comment (MadLittleMods) [reply synapse/appservice/__init__.py:187]: [MSC3905](https://github.com/matrix-org/matrix-spec-proposals/pull/3905) is merged and this behavior is ok to change now ✅

File: synapse/appservice/__init__.py [hunk 10]
```diff
@@ -172,12 +172,24 @@ async def _matches_user_in_member_list(
...
+        #
+        # In the future, we can consider re-using
+        # `store.get_app_service_users_in_room` which is very similar to this
+        # function but has a slightly worse performance than this because we
+        # have an early escape-hatch if we find a single user that the
+        # appservice is interested in. The juice would be worth the squeeze if
+        # `store.get_app_service_users_in_room` was used in more places besides
+        # an experimental MSC. But for now we can avoid doing more work and
```
Comment (anoadragon453) [inline synapse/appservice/__init__.py:185]: [MSC3905](https://github.com/matrix-org/matrix-spec-proposals/pull/3905) is no longer experimental.

File: synapse/appservice/__init__.py (hunk 9 above)
Comment (anoadragon453) [reply synapse/appservice/__init__.py:177]: I wonder if it's worth adding an entry to the [upgrade notes](https://matrix-org.github.io/synapse/latest/upgrade)?

File: synapse/appservice/__init__.py (hunk 10 above)
Comment (MadLittleMods) [reply synapse/appservice/__init__.py:185]: This is talking about `store.get_app_service_users_in_room` usage which is only used by `_compute_msc3202_otk_counts_and_fallback_keys`

https://github.com/matrix-org/synapse/blob/d902181de98399d90c46c4e4e2cf631064757941/synapse/appservice/scheduler.py#L293-L330

File: tests/handlers/test_appservice.py [hunk 11]
```diff
@@ -416,6 +417,157 @@ def prepare(self, reactor, clock, hs):
...
+            self.assertEqual(events[0]["type"], "m.room.message")
+            self.assertEqual(events[0]["sender"], alice)
+        else:
+            self.send_mock.assert_not_called()
+
+    def test_application_services_receive_events_sent_by_interesting_local_user(self):
+        """
+        Test to make sure that a messages sent from a local user can be interesting and
+        picked up by the appservice.
+        """
```
Comment (MadLittleMods) [inline tests/handlers/test_appservice.py:519]: > I specifically wanted to make sure that an appservice registering interest in `*` users would still receive an event that was sent from a local user to a remote user. As in, the appservice would still receive events *sent* from an interesting user.
> 
> And here I believe that is the case, though it may be nice to check `send_mock` was called or not called both when a message is sent from alice *and* the parameterized user. Just to make sure.
>
> *-- @anoadragon453, https://github.com/matrix-org/synapse/pull/13958#pullrequestreview-1154474364*

Added a specific test for this ✅

File: synapse/appservice/__init__.py (hunk 9 above)
Comment (MadLittleMods) [reply synapse/appservice/__init__.py:177]: Added something to the upgrade notes 👍

File: docs/upgrade.md [hunk 12]
```diff
@@ -88,6 +88,28 @@ process, for example:
...
+
+## Changes to the events received by application services (interest)
+
+To align with spec (changed in
+[MSC3905](https://github.com/matrix-org/matrix-spec-proposals/pull/3905)), Synapse now
+only considers local users to be interesting. In other words, the `users` namespace
+regex is only be applied against local users of the homeserver.
+
+Please note, this probably doesn't affect the expected behavior of your application
+service since an interesting local user in a room, still means all messages in the room
```
Comment (anoadragon453) [inline docs/upgrade.md:102]:
```suggestion
service, since an interesting local user in a room still means all messages in the room
```

**Review by anoadragon453 — APPROVED** [review]
LGTM!

**Comment by MadLittleMods** [comment]
Thanks for the review @squahtx, @clokep, @anoadragon453, and @turt2live 🦨
